Microsoft.Research/Contracts/System/System.Collections.ObjectModel.ObservableCollection.cs

```diff
@@ -0,0 +1,193 @@
+        //     Typically, this means when there are additional attempts to change this collection
+        //     during a System.Collections.ObjectModel.ObservableCollection<T>.CollectionChanged
+        //     event. However, it depends on when derived classes choose to call System.Collections.ObjectModel.ObservableCollection<T>.BlockReentrancy().
+        //protected void CheckReentrancy();
+        //
+        // Summary:
+        //     Removes all items from the collection.
+        protected override void ClearItems()
+        {
+        }
+
+        //
+        // Summary:
+        //     Inserts an item into the collection at the specified index.
+        //
+        // Parameters:
+        //   index:
+        //     The zero-based index at which item should be inserted.
+        //
+        //   item:
+        //     The object to insert.
+        protected override void InsertItem(int index, T item)
+        {
+        }
+        //
+        // Summary:
+        //     Moves the item at the specified index to a new location in the collection.
+        //
+        // Parameters:
+        //   oldIndex:
+        //     The zero-based index specifying the location of the item to be moved.
+        //
+        //   newIndex:
+        //     The zero-based index specifying the new location of the item.
+        public void Move(int oldIndex, int newIndex)
+        {
+            Contract.Requires(oldIndex < Count);
+            Contract.Requires(newIndex < Count);
+        }
+        //
+        // Summary:
+        //     Moves the item at the specified index to a new location in the collection.
+        //
+        // Parameters:
+        //   oldIndex:
+        //     The zero-based index specifying the location of the item to be moved.
+        //
+        //   newIndex:
+        //     The zero-based index specifying the new location of the item.
+        protected virtual void MoveItem(int oldIndex, int newIndex)
+        {
+            Contract.Requires(oldIndex < Count);
+            Contract.Requires(newIndex < Count);
+        }
+        //
+        // Summary:
+        //     Raises the System.Collections.ObjectModel.ObservableCollection<T>.CollectionChanged
+        //     event with the provided arguments.
+        //
+        // Parameters:
+        //   e:
+        //     Arguments of the event being raised.
+        protected virtual void OnCollectionChanged(NotifyCollectionChangedEventArgs e)
+        {
+            Contract.Requires(e != null);
+        }
+        //
+        // Summary:
+        //     Raises the System.Collections.ObjectModel.ObservableCollection<T>.PropertyChanged
+        //     event with the provided arguments.
+        //
+        // Parameters:
+        //   e:
+        //     Arguments of the event being raised.
+        protected virtual void OnPropertyChanged(PropertyChangedEventArgs e)
+        {
+            Contract.Requires(e != null);
```

Comment:
This makes sense to me but I don't see anything that would indicate it is required. Could you justify with a comment maybe?